Index synthesized wrapper refs by substituting the Clang decl.

dylansturg opened the pull request:
The goal of this change is replace the references to synthesized declarations, whose definitions are never contained in the indexstore, with references to the underlying declarations which are contained in the indexstore. The current behavior prevents code navigation from the references to the declaration.

Reference: https://github.com/swiftlang/swift/issues/79912

- I added the EnumElementDecl generated from the underlying Clang EnumConstantDecl to the imported decl cache for ClangImporter.
- I took the effective Clang node and corresponding imported decl to substitute for "synthesized wrapper decls".

This has a weird outcome, where a reference to the `Code` enum inside of the synthesized struct can emit 2 references to the underlying enum. There's an example in the test case, where `TestError.Code` is referenced.

Fixes #79912

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

dylansturg committed:
Use Clang node USR for decls synthesized for Error enums, and allow requesting the Swift USR.

The Error enum synthesized declarations, e.g. the struct and its static accessors, should generally appear to be identical to the underlying Clang definitions. There are some specific use cases where the synthesized declarations are necessary though.

I've added an option for USR generation to override the Clang node and emit the USR of the synthesized Swift declaration. This is used by SwiftDocSupport so that the USRs of the synthesized declarations are emitted.

Fixes 79912

bnbarham reviewed (commented):
As I was looking at this I was wondering how/if cursor info handles this case. That then brought me to https://github.com/swiftlang/swift/blob/main/lib/AST/USRGeneration.cpp#L187 (added in https://github.com/swiftlang/swift/commit/62f30d50d200ed799ba46f07e35e7e899ed648c2).

We've apparently run into this before, where previously both the static variable and the enumerator had the same USR and that causes other issues (as then there's no way to uniquely identify the struct/enum and property/enumerator). I need to think about this a bit more - the USR returned from cursor info for these has to match the USR we add to the index.

dylansturg commented:
I had also considered emitting index data for the synthesized declarations. It became complex because the Swift indexing components don't emit ideal data from the Module with the Clang decls. For example, file references are disabled when indexing a module without Swift source files.

I think it could make sense to emit the synthesized declarations. I previously enabled that for decls synthesized from Swift (e.g. the synthesized Codable conformance).

bnbarham commented:
I think my preference here would just be to move that USR adjustment logic into `initDocEntityInfo` only. Then everything else should just work the way you'd expect?

dylansturg commented:
> I think my preference here would just be to move that USR adjustment logic into initDocEntityInfo only. 

Is the intent for the USR for the synthesized struct type and its static accessors use the USRs of the Clang enum and enum constants (respectively), for the index data and cursor info? And in the doc info output, the struct and accessors have unique USRs that are different from the underlying Clang enum and enum constants?

That seems feasible but I'm not sure how to move that USR adjustment. After I remove the adjustment logic from `USRGenerationRequest`, then requests will use the underlying Clang node to generate a USR for the static accessors. Which is reasonable when generating index data. When `initDocEntityInfo` needs the Swift USR for the static accessors and wants to ignore the Clang node, does it need to generate the USR in a different way because `USRGenerationRequest` is going to use the underlying Clang node?

dylansturg commented:
> Yeah, for anything that doesn't specifically need the synthesized Swift declarations, which is just documentation I believe (so doc info + symbol graph).

That sounds reasonable to me. I added a flag and, I think, got it working in those places. I noticed that symbol graph wasn't specifically testing for the USR of the synthesized declarations, so I updated the relevant test to check that.

I wrote the `USRGenerationOptions` to try have a single point of documentation for the `emit_synthesized_decls` flag plus allow for some flexibility in the future. But using it in the various `print*USR` APIs proved fairly painful so I just kept the simple bool. I'm open to other approaches there.

@bnbarham Can you take another look and let me know what you think? Thanks!

hamishknight commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -713,20 +713,45 @@ class RequirementRequest :
+  // synthesized from a Clang based declaration. Useful in cases where Swift
+  // declarations are synthesized from Clang nodes but the caller actually
+  // wants the USR of the Swift declaration.
+  bool emit_synthesized_decls;
```
IMO "emit" seems like the wrong word here, maybe it should be something more like `distinguishSynthesizedDecls`?

hamishknight commented on include/swift/AST/USRGeneration.h:
```diff
@@ -44,7 +44,8 @@ bool printDeclTypeUSR(const ValueDecl *D, raw_ostream &OS);
 /// Prints out the USR for the given ValueDecl.
 /// \returns true if it failed, false on success.
-bool printValueDeclUSR(const ValueDecl *D, raw_ostream &OS);
+bool printValueDeclUSR(const ValueDecl *D, raw_ostream &OS,
```
Could you add the doc comment for the parameter here too?

dylansturg replied:
Done.

hamishknight commented on include/swift/AST/USRGeneration.h:
```diff
@@ -65,7 +66,8 @@ bool printExtensionUSR(const ExtensionDecl *ED, raw_ostream &OS);
 /// Prints out the USR for the given Decl.
 /// \returns true if it failed, false on success.
-bool printDeclUSR(const Decl *D, raw_ostream &OS);
+bool printDeclUSR(const Decl *D, raw_ostream &OS,
```
And here

dylansturg replied:
Done.

hamishknight commented:
> I don't have much better of an idea here other than passing a flag through (and then we could keep it there). Pretty ugly though :( Any other ideas @hamishknight?

Yeah, I'm okay with passing a flag through

dylansturg commented:
I think I have everything updated. Please take another look. Thanks!

hamishknight commented on lib/AST/USRGeneration.cpp:
```diff
@@ -184,8 +191,10 @@ swift::USRGenerationRequest::evaluate(Evaluator &evaluator,
-    ClangNode ClangN = D->getClangNode();
+  auto interpretAsClangNode = [&options](const ValueDecl *D) -> ClangNode {
+    auto &ctx = D->getASTContext();
+    auto *importer = static_cast<ClangImporter *>(ctx.getClangModuleLoader());
```
Ah sorry I just noticed this, we can't `static_cast` to `ClangImporter` here because libAST does not depend on libClangImporter (which is why it's exposed as ClangModuleLoader). You'll instead need to add `getEffectiveClangNode` as a virtual member to `ClangModuleLoader`.

dylansturg replied:
Makes sense. Thanks for explaining.

hamishknight commented on lib/AST/USRGeneration.cpp:
```diff
@@ -10,25 +10,26 @@
 #include "swift/AST/TypeCheckRequests.h"
 #include "swift/AST/USRGeneration.h"
 #include "swift/Basic/Assertions.h"
+#include "swift/ClangImporter/ClangModule.h"
```
You'll want to remove this import too

hamishknight reviewed (commented):
One last comment, but otherwise this LGTM, thanks!

dylansturg committed:
fixup! Use Clang node USR for decls synthesized for Error enums, and allow requesting the Swift USR.

QuietMisdreavus commented on test/SymbolGraph/ClangImporter/ErrorEnum.swift:
```diff
@@ -18,6 +18,12 @@
 // but it should be hidden for a similar reason
 // CHECK-NOT: "precise": "s:SYsSHRzSH8RawValueSYRpzrlE4hash4intoys6HasherVz_tF::SYNTHESIZED::c:@E@MyErrorCode"
 
+// MyError
+// CHECK-DAG: "precise": "s:SC11MyErrorCodeLeV"
+
+// MyError.errFirst
+// CHECK-DAG: "precise": "s:SC11MyErrorCodeLeV8errFirstSoAAVvpZ"
```
Thanks for updating this test! This would have been my only concern with this PR, but it looks like that's settled.

bnbarham commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -713,20 +713,45 @@ class RequirementRequest :
 
+struct USRGenerationOptions {
+  /// @brief Whether to emit USRs using the Swift declaration when it is
+  // synthesized from a Clang based declaration. Useful in cases where Swift
+  // declarations are synthesized from Clang nodes but the caller actually
+  // wants the USR of the Swift declaration.
```
```suggestion
  /// synthesized from a Clang based declaration. Useful in cases where Swift
  /// declarations are synthesized from Clang nodes but the caller actually
  /// wants the USR of the Swift declaration.
```

bnbarham reviewed (approved):
One tiny fix-up, otherwise LGTM. Thanks @dylansturg!

dylansturg committed:
Update include/swift/AST/TypeCheckRequests.h

Co-authored-by: Ben Barham <b.n.barham@gmail.com>

hamishknight commented:
@swift-ci please test

drexin commented:
@swift-ci test
